### PR TITLE
Computed Fields

### LINKED ISSUE #935: Serializing @property
# Question

* OS: Linux(Ubuntu 18)
* Python version 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
* Pydantic version 1.0

I've pored over the documentation and issue and can't seem to find an example for serializing members other than class attributes.  

```py
class Rectangle(BaseModel):
    width: int
    length: int

    @property
    def area(self) -> int:
        return self.width * self.length

r = Rectangle(width=10, length=5)

print(r.json())

# What I'd like to see:
{ "width": 10, "length": 5, "area": 50 }
```
I'd also think that properties without setters would default to readOnly on the schema side as well.

It did seem there was a way to get something close to what I'm after via ORM mode and masking a property on the ORM model, but I'd hate to need to create a class just to contain properties.

Going over [BaseModel._calculate_keys](https://github.com/samuelcolvin/pydantic/blob/25f657fd5ff03f661cb84656747b0a90bad02c46/pydantic/main.py#L653) it seems BaseModel will not yield what I'm after.

Still, I'm hoping that it's just been a long day and I overlooked something :)

### LINKED ISSUE #1241: cached_property issues
# Unable to use cached_property

Hi, 

I am using pydantic for almost any project right now and I find it awesome.

I recently found an handy package, [funcy](https://github.com/Suor/funcy), and I am trying to work with `cached_property` decorator.

What I am doing is something like this:

```py
class A(BaseModel):

    class Config:
        arbitrary_types_allowed = True

    @cached_property
    def test(self):
        return 10
```

The `arbitrary_types_allowed = True` is needed because if not present, a `ValidationError` is raised:

```
File "pydantic/main.py", line 215, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 277, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 249, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 335, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 468, in pydantic.fields.ModelField.populate_validators
  File "pydantic/validators.py", line 574, in find_validators
RuntimeError: no validator found for <class 'funcy.objects.cached_property'>, see `arbitrary_types_allowed` in Config
```

When i do:
```py
a = A()

print(a.test)

# this is what i get
<funcy.objects.cached_property object at 0x7f70466e4748>

# while of course i would like to have 10 as an output.
```

Same happens with python3.8 `cached_property`.
If doing the same with a vanilla python class, the behaviour is the expected one (the property outputs 10).

Is there anything I am missing?
Could anyone help me?

Thank you very much, pydantic is really great and you are doing a terrific job.

### LINKED ISSUE #2313: Pydantic tries to validate method decorated with functools.singledispatchmethod as class attribute
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

See code example:
```py
from pydantic import BaseModel
from functools import singledispatchmethod

class A(BaseModel):
    x: int
        
    @singledispatchmethod
    def add_to_x(self, y):
        raise NotImplementedError()
```

Running this leads to:

```
RuntimeError                              Traceback (most recent call last)
<ipython-input-8-f3e84a8e5cb2> in <module>
      2 from functools import singledispatchmethod
      3 
----> 4 class A(BaseModel):
      5     x: int
      6 

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/main.cpython-38-darwin.so in pydantic.main.ModelMetaclass.__new__()

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/fields.cpython-38-darwin.so in pydantic.fields.ModelField.infer()

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/fields.cpython-38-darwin.so in pydantic.fields.ModelField.__init__()

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/fields.cpython-38-darwin.so in pydantic.fields.ModelField.prepare()

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/fields.cpython-38-darwin.so in pydantic.fields.ModelField.populate_validators()

~/.venvs/clarinet/lib/python3.8/site-packages/pydantic/validators.cpython-38-darwin.so in find_validators()

RuntimeError: no validator found for <class 'functools.singledispatchmethod'>, see `arbitrary_types_allowed` in Config
```

Surely this is not intended behaviour? Is it a feature of functools that this property is added as a result of calling `singledispatchmethod`? 

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.3
            pydantic compiled: True
                 install path: /Users/phinate/.venvs/clarinet/lib/python3.8/site-packages/pydantic
               python version: 3.8.3 (default, May 19 2020, 13:54:14)  [Clang 10.0.0 ]
                     platform: macOS-10.15.7-x86_64-i386-64bit
     optional deps. installed: []
```

### LINKED ISSUE #4609: Mistakenly including cached_property as a field
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

Python 3.9.13
pydantic  1.9.1 and 1.10.2 (both version affected)

after first call cached_property creates field in model __dict__, then pydantic starts including this field in .dict() serialization. Insertion always happens, regardless on Config: extra = Extra.forbid/ignore and regardless on property name - property with name __foo still affected.

### Example Code

```Python
from functools import cached_property

from pydantic import __version__, BaseModel

print(__version__)


class A(BaseModel):
    a: int

    class Config:
        keep_untouched = (cached_property,)

    @cached_property
    def foo(self):
        print('calculated')
        return lambda x: self.a # Some unseriliazable result to trigger exception


# Ok case
obj = A(a=1)
print(obj.dict())
# {'a': 1}
print(obj.json())
# "{"a": 1}"

# Exception after obj.foo call
obj = A(a=1)
print(obj.foo) # Internally modifies obj.__dict__
print(obj.dict())
# Expected: {'a': 1}
# Actual: {'a': 1, 'foo': <function A.foo.<locals>.<lambda> at 0x7fa648032160>} # Whoops, foo in dict 🙄🙄🙄
print(obj.json())
# Expected: "{"a": 1}"
# Actual: Exception
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.9.1 
            pydantic compiled: True
                 install path: /home/manakov_dev/.cache/pypoetry/virtualenvs/etl-u0m7Z2iP-py3
.9/lib/python3.9/site-packages/pydantic
               python version: 3.9.13 (main, Aug  8 2022, 20:56:30)  [GCC 9.4.0]
                     platform: Linux-4.4.0-19041-Microsoft-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [X] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by samuelcolvin**
At long last ... the most requested feature in pydantic.

* closes #935
* closes #1241
* closes #2313
* closes #4609
* replaces #2625 - thank you so much @PrettyWood for your initial work, I've copied a lot of your ideas and code here 🙏 

The example from #2625 works unchanged (except `underscore_attrs_are_private` has been removed):

```py
from functools import cached_property
from random import randint

from pydantic import BaseModel, computed_field


class Square(BaseModel):
    side: float
    _private_attr: int = 1

    @computed_field(title='The area')
    @property
    def area(self) -> float:
        return self.side ** 2

    @area.setter
    def area(self, area: float) -> None:
        self.side = area ** .5

    @area.deleter
    def area(self) -> None:
        self.side = 0.0

    @computed_field(alias='The random number')
    @cached_property
    def random_n(self) -> int:
        return randint(0, 1_000)

    @property
    def public_attr(self) -> int:
        return self._private_attr + 1

    @public_attr.setter
    def public_attr(self, v: int) -> None:
        self._private_attr = v - 1


sq = Square(side=10)
assert sq._private_attr == 1
assert sq.public_attr == 2
sq.public_attr = 5
assert sq._private_attr == 4
assert sq.public_attr == 5
the_random_n = sq.random_n
assert sq.model_dump(by_alias=True) == {'side': 10, 'area': 100, 'The random number': the_random_n}
sq.area = 49
assert sq.model_dump() == {'side': 7, 'area': 49, 'random_n': the_random_n}
del sq.area
assert sq.model_dump() == {'side': 0, 'area': 0, 'random_n': the_random_n}
assert sq.model_dump(exclude={'random_n'}) == {'side': 0, 'area': 0}
```

The following changes have been made from #2625:
* computed fields are not include in JSON Schema at the moment - we need #4697 to generate a separate JSON Schema for the output/serialization of a pydantic model - computed fields will be included in that
* `x: int = ComputedField(...)` style usage has been removed - as @dmontagu pointed out, this will confuse IDEs as it looks like the field should be filled at init time
* the return values of computed fields are no longer validated - your a grown-up make sure you return the right type
* Some of the funkier passthrough of attributes to custom properties e.g. the `update_g` test is not long possible - `ComputedFieldInfo` only proxies `setter` and `deleter` to the wrapped property
* `computed_field` on pydantic dataclasses is also supported
* serialisation of computed properties is managed in rust, see https://github.com/pydantic/pydantic-core/pull/547 so should be significantly more performant than #2625 

TODO:
* [x] copy docs across

selected reviewer: @adriangb

### CODE REVIEW

**Comment by samuelcolvin** [comment]
please review.

**Comment by vitalik** [comment]
Hi @samuelcolvin 

Looks great!

Have a few questions:


1 - is `property` mark required ? 

```Python
    @computed_field(title='The area')
    @property
    def area(self) -> float:
        return self.side ** 2

```
or it can be just like this:
```Python
    @computed_field(title='The area')
    def area(self) -> float:
        return self.side ** 2

```


----

2 - can a computed field have access to `context` or `object` that you pass into `model_validate` ?

File: pydantic/_internal/_decorators.py
Comment (dmontagu) [reply pydantic/_internal/_decorators.py:187]: I'm having a hard time understanding what you are referring to in the linked pull request — could you give me a line to search for or similar? I'd at least be interested to understand what was originally done, even if it doesn't make sense to support here

File: pydantic/_internal/_decorators.py
Comment (dmontagu) [reply pydantic/_internal/_decorators.py:169]: Are these currently collected by the `decorators = _decorators.DecoratorInfos.build(cls)` line? I'm just thinking about how we're going to make this work with non-pydantic dataclasses (with AnalyzedType or whatever), seems like the more unified we can make the collection of these decorators, the better.

File: pydantic/_internal/_decorators.py
Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:187]: search for `__getattr__` and `update_g`.

File: pydantic/_internal/_decorators.py
Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:169]: yes, the `_decorators.DecoratorInfos.build` logic is pretty much the same, we just have a different `isinstance` check.

File: pydantic/_internal/_model_construction.py
Comment (dmontagu) [reply pydantic/_internal/_model_construction.py:83]: There is another reference to `all_ignored_types` below, should we make the change there as well? It's only for annotations, which obviously seem less likely to be in `functools`.

I guess you probably already considered this, but anyway, just pointing it out just in case.

Comment (dmontagu) [reply pydantic/_internal/_model_construction.py:83]: (Also, to be clear, I think this change is an improvement.)

Comment (samuelcolvin) [reply pydantic/_internal/_model_construction.py:83]: it didn't break anything, so I was lazy, should probably fix.

File: pydantic/main.py
Comment (dmontagu) [reply pydantic/main.py:279]: Could we do the following:
* Move `ComputedFieldInfo` into `pydantic.fields`
* Make `model_computed_fields` into a `dict[str, ComputedFieldInfo]`
* Store `model_computed_fields` as a class attribute on the model, like we do with 
  ```python
  # from BaseModel: ...
  model_fields: typing.ClassVar[dict[str, FieldInfo]] = {}
  ```
This feels like it would be a lot more similar to the way `model_fields` works, and I noticed in the tests below it looks like every access to an item from this dict is accessing the `info` attribute, which is of type `ComputedFieldInfo`.

If someone really needed access to the `_decorators.Decorator`, they could still access `self.__pydantic_decorators__` (of dubious publicness I guess)

File: pydantic/_internal/_decorators.py
Comment (dmontagu) [reply pydantic/_internal/_decorators.py:187]: @samuelcolvin While I see the use case for that, I think it's probably a good idea not to implement that until we get specific feature requests for it (assuming we haven't already), not least of which because it will be a major pain to document (i.e., how it behaves if you _do_ call a custom method from the wrapped descriptor).

I also feel like once we have this merged, a PR for that functionality would be a lot easier to review..

Comment (dmontagu) [reply pydantic/_internal/_decorators.py:187]: Also, I think users can potentially achieve the same functionality by doing a delayed decoration with the computed field. Namely, I think it might work to do:
```python
    class User(BaseModel):
        first: str
        last: str

        @MyCustomProperty
        def fullname(self) -> str:
            return f'{self.first} {self.last}'

        @fullname.update_g
        def new_fullname(self) -> str:
            return f'The REAL {self.first} {self.last}'

        fullname = computed_property(fullname)
```

File: pydantic/main.py
Comment (samuelcolvin) [reply pydantic/main.py:279]: the main thing you'd want that isn't on `ComputedFieldInfo` is the property itself, which you could easily get with `getattr(model, field_name)`.

I think that's a good idea, the only change I might propose (and we could definitely make this change if/when it's needed), is to store them both in dunder methods `__pydantic_fields__` and `__pydantic_computed_fields__`, then expose them via a property, so they can work on `BareModel` too if we get there.

File: pydantic/_internal/_decorators.py
Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:187]: makes sense, let's leave it for now.

File: pydantic/main.py
Comment (dmontagu) [reply pydantic/main.py:279]: That all sounds good to me.

**Review by dmontagu — APPROVED** [review]
Other than the comments I made (which may or may not merit changes, I'll let you decide) this all looks very good to me

File: pydantic/_internal/_decorators.py
Comment (adriangb) [reply pydantic/_internal/_decorators.py:169]: Why don't we expand/rename `PydanticDecoratorMarker` into `DescriptorProxy`, add all of the relevant proxy methods (it's only 4 I think, maybe 30 LOC) and then use that everywhere? What we really want in all of these cases is for it to function as the original descriptor if accessed directly but hold extra metadata that we need when introspecting the model.

Something like this should do the trick:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Generic, TypeVar

T = TypeVar('T')


@dataclass
class DescriptorProxy(Generic[T]):
    wrapped: Any
    metadata: T

    def __post_init__(self) -> None:
        descriptor_methods = ('__get__', '__set__', '__delete__', '__set_name__')
        for method in descriptor_methods:
            if hasattr(self.wrapped, method):
                setattr(self, method, getattr(self.wrapped, method))
```

Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:169]: good idea, I'll look at doing that.

**Comment by davidfant** [comment]
Just wanted to say that I was _just_ looking for this and was very pleasantly surprised that this was literally being done as I was searching for a solution. Great work! 🙌

File: pydantic/_internal/_decorators.py
Comment (PrettyWood) [reply pydantic/_internal/_decorators.py:187]: Yup I probably went a bit far 😅

File: pydantic/main.py
Comment (samuelcolvin) [reply pydantic/main.py:279]: done, but kept `model_computed_fields` as a property.

File: pydantic/_internal/_decorators.py
Comment (samuelcolvin) [reply pydantic/_internal/_decorators.py:169]: done, although the weirdness of how `__set_name__` is used means I have to have `PydanticDescriptorProxy` and `PydanticFullDescriptorProxy` which is a subclass which includes `__set_name__` etc.

Happy to change names and have a single class if you can get it to work simply.

**Comment by samuelcolvin** [comment]
As far as I'm concerned, this is ready to merge.

**Comment by samuelcolvin** [comment]
Don't merge, there's some missing code.

**Comment by caniko** [comment]
I wish @samuelcolvin would've referred to #4650, which of the points in the main post have been implemented?

I f not all, any roadmap for the lacking features?

**Comment by samuelcolvin** [comment]
Sorry, I didn't see that issue, I did look, but as you say, there's a lot of noise.

Inheritance is tested and works fine.

Frozen should work fine, needs a test, also a test for validate assignment.

Classmethod - I'm not clear what your asking for, could you give an example?

`abstractmethod` should work, bit needs a test.

I'm about to get on a flight to pycon, I'll check those on the plane or when I land.

**Comment by caniko** [comment]
> Classmethod - I'm not clear what your asking for, could you give an example?

```python
@classmethod
@property
def some_class_property(cls):
    ...

@classmethod
@cached_property
def some_class_property(cls):
    ...
```

Wish you a safe flight :superhero:

**Comment by samuelcolvin** [comment]
flight delayed ... everything works fine except there's not `TypeError` on missing `abstractmethod`s - see the `xfail`.

Question @adriangb @dmontagu @caniko:
* Am I doing the abstractmethod tests right?
* Do we care about supporting this?

**Comment by caniko** [comment]
I am invariant when it comes to `abstractmethod`, here is the [link](https://github.com/pydantic/pydantic/pull/2625#issuecomment-1150460832) to the discussion in the previous.

I would use `computed_fields` with abstract methods that are properties if it is idiomatic. Is it?

**Comment by samuelcolvin** [comment]
If we can make the TypeError work as if there's no `comouted_field` decorator, sounds like a good idea.

Anyone know how to do that, it's not obvious

File: tests/test_computed_fields.py
Comment (adriangb) [inline tests/test_computed_fields.py:461]: I'm not sure what the real-world use case for this is, but this looks good. Do we support or want to support `@abstractmethod` with validators and other decorators we provide?

File: tests/test_computed_fields.py
Comment (caniko) [inline tests/test_computed_fields.py:472]: This can be fixed by making the `computed_field` decorator check if `__isabstractmethod__` is True on the method it is being applied to. If it is True, make sure the respective computed field attribute has the attribute `__isabstractmethod__` set to True.

File: tests/test_computed_fields.py
Comment (caniko) [reply tests/test_computed_fields.py:461]: I already gave my arguments in the main thread, [this](https://github.com/pydantic/pydantic/pull/2625#issuecomment-1124803384) is from the previous PR.

File: tests/test_computed_fields.py
Comment (adriangb) [inline tests/test_computed_fields.py:468]: Why do they have to raise an error? `@abstractmethod` on its own doesn't do that. The class also needs to inherit from `abc.ABC` (runtime checked) or typing.Protocol (static type checker enforced).

Comment (adriangb) [reply tests/test_computed_fields.py:468]: Here's what I suggest should work:

```python
from abc import ABC, abstractmethod

import pytest

from pydantic import BaseModel, computed_field


class AbstractSquare(BaseModel, ABC):
    @computed_field
    @property
    @abstractmethod
    def area(self) -> float:
        raise NotImplementedError

class BadSquare(AbstractSquare):
    pass

class GoodSquare(AbstractSquare):
    @computed_field
    @property
    def area(self) -> float:
        return 1

assert GoodSquare().area == 1

with pytest.raises(TypeError):
    BadSquare()
```

Comment (adriangb) [reply tests/test_computed_fields.py:468]: Or similar. Point is, we don't necessarily want to give a TypeError unless ABC is also in the mix. Might be complicated w/ metaclasses and such.

File: tests/test_computed_fields.py
Comment (caniko) [reply tests/test_computed_fields.py:472]: Implement @adriangb suggestion, and check if it is still not working before going down this router.

**Comment by pawel-swiecki-saucelabs** [comment]
> I would use `computed_fields` with abstract methods that are properties if it is idiomatic. Is it?

**`property`** is a method that is disguised as a field. Idiomatic/textbook example:

```python
class Person:
    @property
    def age(self) -> int:
        # compute age based on `self.birth_date` and `now()`
        pass
```

**`abstractmethod`** is an ABC method that needs to be implemented in non-ABC child classes. Idiomatic/textbook example:

```python
class EventProducer(ABC):
    @abstractmethod
    def produce(self):
        pass
```

Now let's combine the two, similarly to as in @adriangb's [code](https://github.com/pydantic/pydantic/pull/5502#discussion_r1171216151) (modified it a bit):

```python
class Shape2D:
    @property
    @abstractmethod
    def area(self) -> int:
        pass
```

I don't know what to say more than it looks idiomatic.

Now, if this is indeed idiomatic then adding `computed_field` on top, should also be, in my opinion.

(I'm not saying this should be done in this MR. Maybe releasing it without it and waiting for feedback will be a better approach?)

**Comment by dmontagu** [comment]
Regarding how to get the type error, I was able to produce an error in Adrian's example via:

```python
from abc import ABC, abstractmethod

from pydantic import computed_field, BaseModel


class AbstractSquare(BaseModel, ABC):
    @computed_field
    @property
    @abstractmethod
    def area(self) -> float:
        raise NotImplementedError


AbstractSquare.__abstractmethods__ = frozenset(set(AbstractSquare.__abstractmethods__) | {'area'})


class BadSquare(AbstractSquare):
    pass


BadSquare()
#> TypeError: Can't instantiate abstract class BadSquare with abstract method area
```
So I think if we manually update `__abstractmethods__` in the `Metaclass.__new__` based on the computed fields it should work. I think if we just store whether the wrapped property/method is abstract it should be possible to store the relevant info then update the `__abstractmethods__` attribute as appropriate.

And better yet I think we could do this for all the decorators, including `@field_validator` etc. in the same manner.

**Comment by caniko** [comment]
@dmontagu, this can be easily implemented on the `computed_fields` side, why not finish it off now that it is apparent.

Using `@abstractmethod` on a method sets the method's `__isabstractmethod__` attribute to True. Just do the same when applying `@computed_field` to a method with `__isabstractmethod__` set to True.

File: tests/test_computed_fields.py
Comment (samuelcolvin) [reply tests/test_computed_fields.py:472]: As I've said a few times in this thread, dynamically setting these methods doesn't work, at least for `__set_name__`.

Comment (adriangb) [reply tests/test_computed_fields.py:478]: I don't know where you sadi that but yes you are right, to my surprise that doesn't work. Not even overriding `__getattr__` works for those methods. It does work for `__isabstractmethod__` and I just pushed that.

File: tests/test_computed_fields.py
Comment (dmontagu) [inline tests/test_computed_fields.py:487]:
```suggestion
    with pytest.raises(TypeError, match="Can't instantiate abstract class Square with abstract methods? area"):
```

**Comment by adriangb** [comment]
Merging this since it seems to be 99% ready. We can always iterate on that last 1% or fix any issues in follow-ups.

**Pull request merged**
